crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,411 @@
+                    bcs::to_bytes(&transaction.effects)?,
+                ),
+                (EVENTS_COLUMN_QUALIFIER, bcs::to_bytes(&transaction.events)?),
+                (
+                    TIMESTAMP_COLUMN_QUALIFIER,
+                    bcs::to_bytes(&transaction.timestamp)?,
+                ),
+                (
+                    CHECKPOINT_NUMBER_COLUMN_QUALIFIER,
+                    bcs::to_bytes(&transaction.checkpoint_number)?,
+                ),
+            ];
+            items.push((transaction.transaction.digest().inner().to_vec(), cells));
+        }
+        self.multi_set(TRANSACTIONS_TABLE, items).await
+    }
+
+    async fn save_checkpoint(&mut self, checkpoint: &CheckpointData) -> Result<()> {
+        let summary = &checkpoint.checkpoint_summary;
+        let contents = &checkpoint.checkpoint_contents;
+        let key = summary.sequence_number.to_be_bytes().to_vec();
+        let cells = vec![
+            (CHECKPOINT_SUMMARY_COLUMN_QUALIFIER, bcs::to_bytes(summary)?),
+            (
+                CHECKPOINT_CONTENTS_COLUMN_QUALIFIER,
+                bcs::to_bytes(contents)?,
+            ),
+        ];
+        self.multi_set(CHECKPOINTS_TABLE, [(key, cells)]).await
+    }
+}
+
+#[async_trait]
+impl KeyValueStoreReader for BigTableClient {
+    async fn get_objects(&mut self, object_keys: &[ObjectKey]) -> Result<Vec<Object>> {
+        let keys: Result<_, _> = object_keys.iter().map(Self::raw_object_key).collect();
+        let mut objects = vec![];
+        for row in self.multi_get(OBJECTS_TABLE, keys?).await? {
+            for (_, value) in row {
+                objects.push(bcs::from_bytes(&value)?);
+            }
+        }
+        Ok(objects)
+    }
+
+    async fn get_transactions(
+        &mut self,
+        transactions: &[TransactionDigest],
+    ) -> Result<Vec<TransactionData>> {
+        let keys = transactions.iter().map(|tx| tx.inner().to_vec()).collect();
+        let mut result = vec![];
+        for row in self.multi_get(TRANSACTIONS_TABLE, keys).await? {
+            let mut transaction = None;
+            let mut effects = None;
+            let mut events = None;
+            let mut timestamp = 0;
+            let mut checkpoint_number = 0;
+
+            for (column, value) in row {
+                match std::str::from_utf8(&column)? {
+                    TRANSACTION_COLUMN_QUALIFIER => transaction = Some(bcs::from_bytes(&value)?),
+                    EFFECTS_COLUMN_QUALIFIER => effects = Some(bcs::from_bytes(&value)?),
+                    EVENTS_COLUMN_QUALIFIER => events = Some(bcs::from_bytes(&value)?),
+                    TIMESTAMP_COLUMN_QUALIFIER => timestamp = bcs::from_bytes(&value)?,
+                    CHECKPOINT_NUMBER_COLUMN_QUALIFIER => {
+                        checkpoint_number = bcs::from_bytes(&value)?
+                    }
+                    _ => error!("unexpected column {:?} in transactions table", column),
```

Comment:
These match statements likely don't work as you can't match on a global const so iirc everything will just match the first match arm. If we want to do this we can:
- explicitly match against the strings themselves
- or, more cleanly we could introduce an enum for each column and we can take the column bytes and do a conversation from bytes into the enum and then match on the enum.

Same goes for the checkpoints below

Reply:
hm, I just verified it locally with an emulator. It works

Reply:
i swear i had to fix something like this recently...well if it works it works!